### ISSUE TITLE
Setting large value for CLOB var fails with cx_Oracle.DatabaseError: ORA-03146: Invalid buffer length for TTC field

### ISSUE THREAD

**Issue by mkmoisen** [oracle/python-cx_Oracle#112]
I'm attempting to set a large value (>536MB) into a cx_oracle CLOB variable, but it appears to be failing around the 512MB mark. I've tested this on two environments with identical results: Python on my local environment connecting to my local Oracle; and python on a remote unix environment connecting to a different remote Oracle environment.

    conn = cx_Oracle.connect('foo/bar@baz')
    cur = conn.cursor()
    clob = cur.var(cx_Oracle.CLOB)

    # This first command is successful
    clob.setvalue(0, 'A' * 1024 * 1024 * 511)

    # This second command fails
    clob.setvalue(0, 'A' * 1024 * 1024 * 512)
    
The second command (or any value greater than this) fails with:

    cx_Oracle.DatabaseError: ORA-03146: Invalid buffer length for TTC field

After making exactly three calls to the second command, the session itself appears to be disconnected (whereas this does not happen with the first command)

    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    cx_Oracle.OperationalError: ORA-03135: connection lost contact
    Process ID: 17567
    Session ID: 137 Serial number: 9226

----
Database version: Oracle 12CR1

cx_Oracle version: Tested on 5.1.3 and 6.0.3 with identical results

Python version: 2.7.13

Everything is 64 bit.

Note: I cross posted this to [stackoverflow](https://stackoverflow.com/questions/47276246/cx-oracle-setting-large-value-for-clob-var-fails-with-cx-oracle-databaseerror-o).

----

#### Answer the following questions:

1. What is your version of Python? Is it 32-bit or 64-bit?
2.7.13 64-bit

2. What is your version of cx_Oracle?
Tested on 5.1.3 and 6.0.3 with identical results

3. What is your version of the Oracle client (e.g. Instant Client)?  How was it
   installed?  Where is it installed?
12.1.0.2.0
On my windows environment, I installed it from Oracles website. On my unix environment, it was installed by the sysadmins, probably with yum.

4. What is your version of the Oracle Database?
12CR1

5. What is your OS and version?
On my Windows, it is Windows 10. On my unix, it is Red Hat Enterprise Linux Server release 6.9 (Santiago)

6. What compiler version did you use?  For example, with GCC, run
   `gcc --version`.
I installed from pip on both windows and unix.

7. What environment variables did you set?  How *exactly* did you set them?
N/A

8. What *exact* command caused the problem (e.g. what command did you try to
   install with)?  Who were you logged in as?

    clob = cur.var(cx_Oracle.CLOB)
    clob.setvalue(0, 'A' * 1024 * 1024 * 512)

9. What error(s) you are seeing?

    cx_Oracle.DatabaseError: ORA-03146: Invalid buffer length for TTC field

**Comment by anthony-tuininga** [comment]
I can replicate this with 12.2 client and cx_Oracle 6. This may be a limitation of the Oracle client. I'll ask internally.

**Comment by anthony-tuininga** [comment]
Do you have to use a temporary LOB? Can you use a permanent LOB instead? It definitely doesn't have these limitations.

**Comment by mkmoisen** [comment]
Hi @anthony-tuininga 

I'm not familiar with temporary vs permanent/persistant LOBs, but I think you are suggesting I use a direct DML insert, correct? We are following a set of ERP standards which mandate that all DMLs must occur through corresponding PLSQL packages only, without exception. This is a hard rule that I cannot break.

Is there any way to leverage a permanent/persistence LOB in cx_oracle when calling `cur.callproc` ?

Here is an example of my intended use case:

    data = get_user_provided_data()
    user = get_user_name()
    
    conn = cx_Oracle.connect('foo/bar@baz')
    cur = conn.cursor()
    clob = cur.var(cx_Oracle.CLOB)
    clob.setvalue(0, data)
    
    cur.callproc('xxfoo_bar.insert_one', 
                       keywordParameters={
                           'p_data_i': clob, 
                           'p_who_i': user
                       })

E.g, find some relatively large data, and call a PLSQL procedure to insert that data as a CLOB.

Thanks and best regards,

Matthew Moisen

**Comment by anthony-tuininga** [comment]
You should be able to call the stored procedure with the CLOB as an OUT variable. The CLOB would be a simple empty_clob() which the Python program could then write to directly. Once the write is complete you would have to make sure that a commit took place. Would that work for your situation?

**Comment by mkmoisen** [comment]
Hi @anthony-tuininga 

I'm not sure I'm following. Would you mind providing an example PLSQL and python code if you have a chance?

**Comment by anthony-tuininga** [comment]
Sure. Something like this:

```sql
create table Test1 (
    Id number(9) not null,
    ClobVal clob not null
);

create or replace package pkg_Test1 as

    procedure insert_one (
        a_Id                number,
        a_CLOB              out clob
    );

end;
/

create or replace package body pkg_Test1 as

    procedure insert_one (
        a_Id                number,
        a_CLOB              out clob
    ) is
    begin
        insert into Test1 (Id, ClobVal)
        values (a_Id, empty_clob())
        returning ClobVal into a_CLOB;
    end;

end;
/
```
And here is the corresponding Python code:
```python
var = cursor.var(cx_Oracle.CLOB)
cursor.callproc("pkg_Test1.insert_one", (1, var))
lob = var.getvalue()
lob.write(buf)
conn.commit()
```
I ran this code and was able to successfully insert 536,870,912 characters into the CLOB.

**Comment by mkmoisen** [comment]
@anthony-tuininga 

Thanks, this appears to be working. There is one issue when attempting to close the connection:

    >>> cursor.close()
    >>> conn.close()
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    cx_Oracle.DatabaseError: DPI-1054: connection cannot be closed when open statements or LOBs exist

I noticed that `lob` has a `close` method, but when I attempt to call it (either before or after attempting to `conn.close()`:

    >>> lob.isopen()
    False
    >>> lob.close()
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    cx_Oracle.DatabaseError: ORA-22289: cannot perform  operation on an unopened file or LOB

Also, would you be able to explain what is going on here - how is `lob.write` able to actually write to the column in the table without us calling another DML/PLSQL?

**Comment by anthony-tuininga** [comment]
First, lob.close() doesn't do what you think it does. Take a look at the [documentation](http://cx-oracle.readthedocs.io/en/latest/lob.html#LOB.close). You need to remove all references to it in order to prevent DPI-1054. This is generally done with the statement "del lob". You can also just let the connection, cursor and LOB close on their own -- which will happen automatically when they go out of scope.

Second, the LOB value that is returned is a "locator" and it retains its connection to the database and where it is actually found in that database. That allows you to write to it without having to perform additional DML. Hope that explanation helps!

**Comment by mkmoisen** [comment]
@anthony-tuininga 

Thanks, that makes sense.

Unfortunately, this workaround won't work in my particular case. Our coding standards require that any client (e.g., Python) connect to a schema that contains no write grants to the schema in which the tables are defined. Rather, the client schema has execute grants on PLSQL packages defined in a different schema which has the write grants to the tables. 

I tried hooking this workaround up into our code base, but it fails with `DatabaseError: ORA-01031: insufficient privileges`, as I'm connecting from python in the client schema, and presumably the CLOB locator requires the appropriate write privileges.

This schema setup is a hard requirement, unfortunately, that I cannot violate.

Do you see another way around this?

Here is the schema standards we use:

* XXFOO - All Tables are stored here. GRANTS for SELECT/DMLS are given to XXFOO_APPS with grant option.
* XXFOO_APPS - All PLSQLs are stored here. Select grants are given to xxfoo_u on the tables in XXFOO, but not DML grants.
* XXFOO_U - All Clients (e.g. python) connect here and call PLSQL packages stores in XXFOO_APPS. This schema does not have DML access to the tables in XXFOO.

**Comment by anthony-tuininga** [comment]
Yes, in order to write to the LOB you need to have update on the column in question. The only other way to manage it that I can think of is to write multiple LOBs, but I'm not sure how palatable that option is either!

**Comment by cjbj** [comment]
@mkmoisen We sent this over to some (other) Oracle developers to look at.  They are using Oracle bug 27147549 to track it.

**Comment by anthony-tuininga** [comment]
Good news! We found a workaround for this issue. There is a limit for each piece of data that you send, but there is a way to write this piecewise. See the following code. Note that this is *substantially* slower than using a permanent LOB as shown above, but it does work!

```python
var = cursor.var(cx_Oracle.CLOB)
var.setvalue(0, "X")    # write a small value first to force the temporary LOB to be created
lob = var.getvalue()    # get the temporary LOB
bufSize = 1024 * 1024 * 513
buf = 'A' * bufSize
print("Length of buf is", len(buf))
splitPos = 1024 * 1024 * 256
lob.write(buf[:splitPos])  # write in portions < 512 MB
lob.write(buf[splitPos:], splitPos + 1)
cursor.execute("insert into Test1 values (1, :val)", val = var)  # calling stored procedure should work, too
conn.commit()
```

**Comment by anthony-tuininga** [comment]
I'm going to close this since a suitable workaround has been found. Feel free to re-open if needed.

**Comment by cjbj** [comment]
General note: if the workaround isn't suitable, contact Oracle Support for a backport of Oracle bug 27147549.
